app/DynamicGraphs/GraphGenerator.hs

```diff
@@ -7,34 +7,110 @@ module DynamicGraphs.GraphGenerator
+-- use reqToStmts to converts a list of requirements all at once and concatenate the results into a 
+-- single list of DotGraph objects.
 reqsToGraph :: [(Text, Req)] -> DotGraph Text
 reqsToGraph reqs =
-    let stmts = concatMap reqToStmts reqs
+    let stmts = stmtslst (foldUpReqLst reqs 0)
     in
         buildGraph stmts
 
-reqToStmts :: (Text, Req) -> [DotStatement Text]
-reqToStmts (name, NONE) = [DN $ DotNode name []]
-reqToStmts (_, _) = []
+-- convert the original requirement data into dot statements that can be used by buildGraph to create the
+-- corresponding DotGraph objects. ([] is the [] after name the optional parameters for DotNode?)
+reqToStmts :: (Text, Req) -> StmtsWithCounter -> StmtsWithCounter
+
+
+reqToStmts (name, NONE) stmtcounter = StmtsWithCounter{stmtslst = connectLst (stmtslst stmtcounter) [DN $ DotNode (name `mappend` (pack (show (counter stmtcounter)))) []]
+                                      , counter = (counter stmtcounter) + 1}
+
+
+
+
+reqToStmts (name, J string1) stmtcounter = StmtsWithCounter{stmtslst = connectLst [DE $ DotEdge str1 name []] (stmtslst (foldUpReqLst [(name, NONE), (str1, NONE)] 0))
+                                                           , counter = (counter stmtcounter) + 3}
+  where str1 = pack string1
+
+--reqToStmts (name, AND reqs1) stmtcounter = connectLst [DN $ DotNode "and" [], DE $ DotEdge "and" name []] (connectLst andedges nodestatements)
+--  where andedges = createAndEdge (dragNodeIdInReq reqs1)
+--       nodestatements = concatMap reqToStmts ((name, NONE):(decompJString reqs1))
+
+--reqToStmts (name, OR reqs1) stmtcounter = connectLst [DN $ DotNode "or" [], DE $ DotEdge "or" name []] (connectLst oredges nodestatements)
+--  where oredges = createOrEdge (dragNodeIdInReq reqs1)
+--        nodestatements = concatMap reqToStmts ((name, NONE):(decompJString reqs1))
+
+--reqToStmts (name, FROM string1 (J string2)) stmtcounter = connectLst fromedges (concatMap reqToStmts [(name, NONE), (pack_string1, NONE), (pack_string2, NONE)])
+--  where pack_string1 = pack string1
+--        pack_string2 = pack string2
+--        fromedges = [DE $ DotEdge pack_string2 pack_string1 [], DE $ DotEdge pack_string1 name []]
+
+--reqToStmts (name, GRADE string1 (J string2)) stmtcounter = connectLst gradeedges (concatMap reqToStmts [(name, NONE), (pack string1, NONE), (pack string2, NONE)])
+--  where pack_string1 = pack string1
+--        pack_string2 = pack string2
+--        gradeedges = [DE $ DotEdge pack_string2 pack_string1 [], DE $ DotEdge pack_string1 name []]
+
+--reqToStmts (name, RAW string1) stmtcounter = connectLst rawedges (concatMap reqToStmts [(name, NONE), (pack_string1, NONE)])
+--  where pack_string1 = pack string1
+--        rawedges = [DE $ DotEdge pack_string1 name []]
+
+--reqToStmts (_, _) stmtcounter = []
+
+data StmtsWithCounter = StmtsWithCounter{ stmtslst :: [DotStatement Text]
```

Comment:
I believe I mentioned this in your last presentation, but your code will be a lot simpler if you use a type synonym like `type StmtsWithCounter = ([DotStatement Text], Int)`. You'll be able to use short pattern matching to extract the two parts of the tuple to use in your code.

Don't worry about extensibility for now -- after this is up and running, we can use some more advanced techniques to carry around mutable state later.

Reply:
Oh, and also you'll want to put the definition of this type near the top of the file, and certainly above `reqToStmts`.

Reply:
Here, I find it kind of difficult in practice since in many cases, I have to visit the attributes of StmtsWithCounter by its name (like stmtslst or counter), if using short form, it becomes hard or at least very inconvenient to visit them. 

For instance, in line 34 "let stmts = stmtslst (foldUpReqLst reqs 0)", I have to visit the stmtslst attribute of the StmtsWithCounter returned by foldUpReqLst, and I do not know how to use pattern match here. And similar problems in other parts.

Reply:
You can pattern-match on tuples in `let` statements as well. It's actually easier to do so than using records and function calls. For example, you could do

```
let stmts, _ = foldUpReqLst reqs 0
```

if you only wanted the first part of the tuple.